client/session.go

```diff
@@ -0,0 +1,277 @@
+	queues    []hostQueue
+
+	// NB(r): we make sets of pools, one for each shard, to reduce
+	// contention on the pool when retrieving ops
+	writeOpPools []writeOpPool
+}
+
+func newSession(opts m3db.ClientOptions) (m3db.Session, error) {
+	topology, err := opts.GetTopologyType().Create()
+	if err != nil {
+		return nil, err
+	}
+
+	s := &session{
+		opts:      opts,
+		nowFn:     opts.GetNowFn(),
+		topo:      topology,
+		topoMapCh: make(chan m3db.TopologyMap),
+	}
+
+	currTopologyMap := topology.GetAndSubscribe(s.topoMapCh)
+	if err := s.setTopologyMap(currTopologyMap); err != nil {
+		return nil, err
+	}
+
+	go func() {
+		log := opts.GetLogger()
+		for value := range s.topoMapCh {
+			if err := s.setTopologyMap(value); err != nil {
+				log.Errorf("could not update topology map: %v", err)
+			}
+		}
+	}()
+
+	return s, nil
+}
+
+func (s *session) setTopologyMap(topologyMap m3db.TopologyMap) error {
+	// NB(r): we leave existing writes in the host queues to finish
+	// as they are already enroute to their destination, this is ok
+	// as part of adding a host is to add another replica for the
+	// shard set and only once bootstrapped decomission the old node
+	start := s.nowFn()
+
+	hosts := topologyMap.Hosts()
+	queues := make([]hostQueue, len(hosts))
+	for i := range queues {
+		queues[i] = s.newHostQueue(hosts[i], topologyMap)
+	}
+
+	shards := topologyMap.ShardScheme().All().Shards()
+	minConnectionCount := s.opts.GetMinConnectionCount()
+	for {
+		if s.nowFn().Sub(start) >= s.opts.GetClusterConnectTimeout() {
+			return ErrClusterConnectTimeout
+		}
+		// Be optimistic
+		clusterHasMinConnectionsToAllShards := true
+		for _, shard := range shards {
+			// Be optimistic
+			shardHasMinConnectionsToOneHost := true
+			routeErr := topologyMap.RouteShardForEach(shard, func(idx int, host m3db.Host) {
+				if queues[idx].GetConnectionCount() < minConnectionCount {
+					shardHasMinConnectionsToOneHost = false
+				}
+			})
+			if routeErr != nil {
+				return routeErr
+			}
+			if !shardHasMinConnectionsToOneHost {
+				clusterHasMinConnectionsToAllShards = false
+				break
+			}
+		}
+		if clusterHasMinConnectionsToAllShards {
+			// All done
+			break
+		}
+		time.Sleep(clusterConnectWaitInterval)
+	}
+
+	s.Lock()
+	prev := s.queues
+	s.queues = queues
+	s.setWriteOpPoolsWithLock(topologyMap)
+	s.topoMap = topologyMap
+	s.Unlock()
+
+	// Asynchronously close the previous set of host queues
+	go func() {
+		for i := range prev {
+			prev[i].Close()
+		}
+	}()
+
+	return nil
+}
+
+func (s *session) newHostQueue(host m3db.Host, topologyMap m3db.TopologyMap) hostQueue {
+	// NB(r): Due to hosts being replicas we have:
+	// = replica * numWrites
+	// = total writes to all hosts
+	// We need to pool:
+	// = replica * (numWrites / writeBatchSize)
+	// = number of batch request structs to pool
+	totalBatches := topologyMap.Replicas() *
+		int(math.Ceil(float64(s.opts.GetWriteOpPoolSize())/float64(s.opts.GetWriteBatchSize())))
```

Comment:
might make sense to add a comment somewhere indicating `GetWriteOpPoolSize` returns the total write op pool size across all shards without replication.